src/main/scala/temple/DSL/semantics/Validator.scala

```diff
@@ -1,44 +1,81 @@
+import temple.DSL.semantics.Validator._
 import temple.ast.AttributeType._
 import temple.ast.{Metadata, _}
+import temple.builder.project.ProjectConfig
 
 import scala.collection.mutable
 import scala.reflect.{ClassTag, classTag}
 
-private class Validator private (templefile: Templefile) {
-  var errors: mutable.Set[String] = mutable.SortedSet()
+private class Validator(templefile: Templefile) {
+  var errors: mutable.Buffer[String] = mutable.Buffer()
 
-  private lazy val allStructs: Iterable[String] = templefile.services.flatMap {
-    case (name, service) => service.structs.keys
+  private val allStructs: Iterable[String] = templefile.services.flatMap {
+    case _ -> service => service.structs.keys
   }
 
-  private lazy val allServices: Set[String] = templefile.services.keys.toSet
+  private val allServices: Set[String] = templefile.services.keys.toSet
 
-  private lazy val allStructsAndServices: Set[String] = allServices ++ allStructs
+  private val allStructsAndServices: Set[String] = allServices ++ allStructs
 
-  private def validateAttributes(attributes: Map[String, Attribute])(context: SemanticContext): Unit = {
-    attributes.keys.foreach { attributeName =>
-      if (attributeName.headOption.forall(!_.isLower))
-        errors += context.errorMessage(s"Invalid attribute name $attributeName, it must start with a lowercase letter,")
+  private var newServices: Map[String, ServiceBlock] = templefile.services
+
+  private val globalRenaming      = mutable.Map[String, String]()
+  def allGlobalNames: Set[String] = globalRenaming.valuesIterator.toSet ++ allStructsAndServices
+
+  private def validateAttributes(
+    attributes: Map[String, Attribute],
+    context: SemanticContext,
+  ): Map[String, Attribute] = {
+    foreachValueWithContext(attributes, context, validateAttribute)
+
+    // This looks like a map, but it isn’t because the accumulator `newAttributes`’ names are accessed during iteration
+    attributes.foldLeft(attributes.mapFactory[String, Attribute]()) {
+      case (newAttributes, (attributeName, value)) =>
+        if (attributeName.headOption.exists(!_.isLower))
+          errors += context.errorMessage(
+            s"Invalid attribute name $attributeName, it must start with a lowercase letter,",
+          )
+        val newName = dodgeNames(
+          attributeName,
+          templefile.projectName,
+          (newAttributes.keys ++ attributes.keys).toSet - attributeName,
+          decapitalize = true,
+        )(postgresValidator)
+        newAttributes + (newName -> value)
     }
-    attributes.foreachEntry(context(validateAttribute))
   }
 
-  private def validateService(service: ServiceBlock)(context: SemanticContext): Unit = {
-    service.structs.foreachEntry(context(validateStruct))
-    validateAttributes(service.attributes)(context)
-    validateMetadata(service.metadata, service.attributes)(context)
+  private def renameBlock(name: String, block: TempleBlock[_]): Unit = {
```

Comment:
A comment here, explaining the side effect please

Reply:
Sure, coming soon